### ISSUE TITLE
No non-allocating multiplication for `CompositeMap`

### ISSUE THREAD

**Issue by junyuan-chen** [JuliaLinearAlgebra/LinearMaps.jl#202]
It seems that `mul!` involving a `CompositeMap` will allocate an array for each intermediate step of the multiplications along the chain of maps. Am I right that there is no such an implementation readily available that avoids such allocations? In my use case, the map needs to be reapplied for many times.

If I implement an alternative type that avoids such memory allocation (by including pre-allocated arrays inside the composite map), would that be something considered suitable for being included into this package? I could make a PR in a couple of days if this sounds fine.

**Comment by dkarrasch** [comment]
Yes, the current status is that you explicitly hook into the `_compositemul!` machinery, which was introduced for exactly that purpose, when users know that they are dealing with `CompositeMap`s and know an appropriate eltype of the intermediate storage arrays. It is somewhat isolated from the rest of the package because no other map type provides such an opportunity. After all, memory optimization has never been within the scope of this package, because it is in its generic form a very difficult task. Note, however, that `_compositemul!` is covered by tests (and called explicitly), so it is "quasi-exported" and guaranteed to stay.

**Comment by daanhb** [comment]
Might it be possible to "plan" the application of a composite map?
Similarly to `plan_fft` and `plan_fft!`, the plan could take an example vector, from which it could deduce the types of all intermediate vectors.

This might still prove to be difficult, but at least the logic of planning the application of a map can be separated from the map itself.
